### ISSUE TITLE
Install error with Pillow on CentOS 7

### ISSUE THREAD

**Issue by csucat** [python-pillow/Pillow#1683]
My server's OS is **CentOS 7** and the Python version is **2.7.5**, the server comes with the default Pillow version **2.0.0**(installed with **"yum install python-imaging"**). But when I wanted to upgrade the Pillow package using **"pip install --upgrade Pillow"**,  error occured.
Specifically, it gives the following error messages:

>  /usr/bin/ld:/usr/lib/libz.so: file format not recognized; treating as linker script
>  /usr/bin/ld:/usr/lib/libz.so:1: syntax error
>  collect2: error: ld returned 1 exit status
>  error: command 'gcc' failed with exit status 1

Then I also tried to install Pillow from source but without success. First, I installed all the dependencies given by [Pillow installation instruction](http://pillow.readthedocs.org/en/3.0.x/installation.html), then I use **"python setup.py install --user"**, it gave exactly the same error messages.

Anything wrong with my installation methods or my setup or others settings?

**Comment by wiredfool** [comment]
I've never seen anything like that, but it does appear that it's an issue farther down in the stack than Pillow. The error that's getting reported is a linker error, that it doesn't understand the format of the libz shared library. It's possible that you don't have the proper build dependencies, or a good toolchain installed. 

There is a script here: https://github.com/wiredfool/Pillow/blob/depends/depends/fedora_23.sh that installs the compile time dependencies for fedora -- but it's for a current version, not the old Centos that you're running. It may give you a start in terms of dependencies.

**Comment by csucat** [comment]
I have tried to install dependency according to the script: https://github.com/wiredfool/Pillow/blob/depends/depends/fedora_23.sh,but i still can not install pillow, same error message occurs:

>   /usr/bin/ld:/usr/lib/libz.so: file format not recognized; treating as linker script
    /usr/bin/ld:/usr/lib/libz.so:1: syntax error
    collect2: error: ld returned 1 exit status
    error: command 'gcc' failed with exit status 1

 > ----------------------------------------
 > Rolling back uninstall of Pillow
Command "/usr/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-build-RatY10/Pillow/setup.py';exec(compile(getattr(tokenize, 'open', open)(__file__).read().replace('\r\n', '\n'), __file__, 'exec'))" install --record /tmp/pip-pSyhxm-record/install-record.txt --single-version-externally-managed --compile" failed with error code 1 in /tmp/pip-build-RatY10/Pillow/

Any ideas how to fix this?

**Comment by wiredfool** [comment]
What is the output of `ls -l /usr/lib/libz.so`?

**Comment by csucat** [comment]
@wiredfool 
Problem solved!  I will write down the solution in case that anyone else meets the same problem.

First, I tried to use `  ll /usr/lib/libz.so*  `,  and got the following output:
>-rw-r--r--. 1 root root     13 Dec 18 15:41 /usr/lib/libz.so
-rw-r--r--. 2 root root 106903 Dec 18 15:41 /usr/lib/libz.so.1
-rw-r--r--. 1 root root     13 Dec 18 15:41 /usr/lib/libz.so1.1
-rw-r--r--. 2 root root 106903 Dec 18 15:41 /usr/lib/libz.so.1.2.5

Then I used `rpm -qf /usr/lib/libz.so`, I got:
>file /usr/lib/libz.so is not owned by any package

The output is similar for other __libz.so.*__  files. So which package does provide the __libz.so*__ files？
 `yum provides *\libz.so` gave the following result:
>zlib-devel-1.2.7-15.el7.i686 : Header files and libraries for Zlib development
Repo        : base
Matched from:
Filename    : /usr/lib/libz.so
zlib-devel-1.2.7-15.el7.x86_64 : Header files and libraries for Zlib development
Repo        : base
Matched from:
Filename    : /usr/lib64/libz.so

Maybe we should use /usr/lib64/libz.so, but somehow gcc tried to use the corrupted or incorrect /usr/lib/libz.so?  In order to verify, I backed up the original /usr/lib/libz.so( `mv /usr/lib/libz.so /usr/lib/libz.so.bak`), and established a symbolic link(`ln -s /usr/lib64/libz.so /usr/lib/libz.so`).

After all these steps, I successfully upgraded Pillow to the latest version using `pip install --upgrade Pillow`, it gave the following message:
>Collecting Pillow
  Using cached Pillow-3.2.0.tar.gz
Installing collected packages: Pillow
  Found existing installation: Pillow 2.0.0
    Uninstalling Pillow-2.0.0:
      Successfully uninstalled Pillow-2.0.0
  Running setup.py install for Pillow ... done
Successfully installed Pillow-3.2.0

Hope it helps~

**Comment by wiredfool** [comment]
Excellent!

I've just tried this in a fresh Centos7 vm and did not have this issue. 

Also, it appears that /usr/lib isn't actually in the library path, so manually installing openjpeg2 to that path works at compile time but fails badly at runtime due to the library not being found.

**Comment by csucat** [comment]
I still do not know why these libz.so* files exist in the directory /usr/lib, they belong to the package **zlib-1.2.7-15.el7.i686**, but I did not install the package. 
Maybe we can change some code to instruct gcc to find the correct libz.so file in the directory /usr/lib64?

**Comment by wiredfool** [comment]
On my centos install, I'm seeing the following paths checked:  (on current master)
```
(vpy27)[centos@centos7 Pillow]$ python setup.py build_ext --debug
Single threaded build, not installing mp_compile:1 processes
running build_ext
Appending path /home/centos/Pillow/libImaging
Appending path /home/centos/vpy27/lib
Appending path /home/centos/vpy27/include
Appending path /usr/lib64
Tkinter found, will check for Tcl/Tk
Checking /py27/Tcl for tk.h
Checking /python27/Tcl for tk.h
Checking /Tcl for tk.h
Checking /Tcl85 for tk.h
Checking /Tcl8.5 for tk.h
Checking /home/centos/Pillow/Tcl for tk.h
Tcl/tk not found
Appending path /usr/local/lib
Appending path /usr/local/include
Appending path /usr/lib
Appending path /usr/include
Looking for zlib
Checking for include file zlib.h in /home/centos/Pillow/libImaging
Checking for include file zlib.h in /home/centos/vpy27/include
Checking for include file zlib.h in /usr/local/include
Checking for include file zlib.h in /usr/include
Found zlib.h
Found library z at /usr/lib64/libz.so
Looking for jpeg
Checking for include file jpeglib.h in /home/centos/Pillow/libImaging
Checking for include file jpeglib.h in /home/centos/vpy27/include
Checking for include file jpeglib.h in /usr/local/include
Checking for include file jpeglib.h in /usr/include
Found jpeglib.h
Found library jpeg at /usr/lib64/libjpeg.so
Looking for jpeg2000
Checking for openjpeg-#.# in /home/centos/Pillow/libImaging
Checking for openjpeg-#.# in /home/centos/vpy27/include
Checking for openjpeg-#.# in /usr/local/include
Checking for openjpeg-#.# in /usr/include
Found openjpeg.h in /usr/include/openjpeg-2.1
Best openjpeg version (2, 1) so far in /usr/include/openjpeg-2.1
Checking for openjpeg-#.# in /usr/include/python2.7
Found library openjp2 at /usr/lib/libopenjp2.so
Inserting path /usr/include/openjpeg-2.1
Looking for tiff
Checking for include file tiff.h in /usr/include/openjpeg-2.1
Checking for include file tiff.h in /home/centos/Pillow/libImaging
Checking for include file tiff.h in /home/centos/vpy27/include
Checking for include file tiff.h in /usr/local/include
Checking for include file tiff.h in /usr/include
Found tiff.h
Found library tiff at /usr/lib64/libtiff.so
Looking for freetype
Found library freetype at /usr/lib64/libfreetype.so
Checking for include file ft2build.h in /usr/include/openjpeg-2.1
Checking for include file ft2build.h in /usr/include/openjpeg-2.1/freetype2
Checking for include file ft2build.h in /home/centos/Pillow/libImaging
Checking for include file ft2build.h in /home/centos/Pillow/libImaging/freetype2
Checking for include file ft2build.h in /home/centos/vpy27/include
Checking for include file ft2build.h in /home/centos/vpy27/include/freetype2
Checking for include file ft2build.h in /usr/local/include
Checking for include file ft2build.h in /usr/local/include/freetype2
Checking for include file ft2build.h in /usr/include
Found ft2build.h in /usr/include
Inserting path /usr/include/freetype2
Looking for lcms
Checking for include file lcms2.h in /usr/include/freetype2
Checking for include file lcms2.h in /usr/include/openjpeg-2.1
Checking for include file lcms2.h in /home/centos/Pillow/libImaging
Checking for include file lcms2.h in /home/centos/vpy27/include
Checking for include file lcms2.h in /usr/local/include
Checking for include file lcms2.h in /usr/include
Found lcms2.h
Found library lcms2 at /usr/lib64/liblcms2.so
Checking for include file tk.h in /usr/include/freetype2
Checking for include file tk.h in /usr/include/openjpeg-2.1
Checking for include file tk.h in /home/centos/Pillow/libImaging
Checking for include file tk.h in /home/centos/vpy27/include
Checking for include file tk.h in /usr/local/include
Checking for include file tk.h in /usr/include
Found tk.h
Looking for TCL
Couldn't find library tcl85 in ['/home/centos/vpy27/lib', '/usr/lib64', '/usr/local/lib', '/usr/lib', '/usr/lib64']
Found library tcl8.5 at /usr/lib64/libtcl8.5.so
Looking for TK
Couldn't find library tk85 in ['/home/centos/vpy27/lib', '/usr/lib64', '/usr/local/lib', '/usr/lib', '/usr/lib64']
Found library tk8.5 at /usr/lib64/libtk8.5.so
Looking for webp
Checking for include file webp/encode.h in /usr/include/freetype2
Checking for include file webp/encode.h in /usr/include/openjpeg-2.1
Checking for include file webp/encode.h in /home/centos/Pillow/libImaging
Checking for include file webp/encode.h in /home/centos/vpy27/include
Checking for include file webp/encode.h in /usr/local/include
Checking for include file webp/encode.h in /usr/include
Found webp/encode.h
Checking for include file webp/decode.h in /usr/include/freetype2
Checking for include file webp/decode.h in /usr/include/openjpeg-2.1
Checking for include file webp/decode.h in /home/centos/Pillow/libImaging
Checking for include file webp/decode.h in /home/centos/vpy27/include
Checking for include file webp/decode.h in /usr/local/include
Checking for include file webp/decode.h in /usr/include
Found webp/decode.h
Found library webp at /usr/lib64/libwebp.so
Looking for webpmux
Checking for include file webp/mux.h in /usr/include/freetype2
Checking for include file webp/mux.h in /usr/include/openjpeg-2.1
Checking for include file webp/mux.h in /home/centos/Pillow/libImaging
Checking for include file webp/mux.h in /home/centos/vpy27/include
Checking for include file webp/mux.h in /usr/local/include
Checking for include file webp/mux.h in /usr/include
Found webp/mux.h
Checking for include file webp/demux.h in /usr/include/freetype2
Checking for include file webp/demux.h in /usr/include/openjpeg-2.1
Checking for include file webp/demux.h in /home/centos/Pillow/libImaging
Checking for include file webp/demux.h in /home/centos/vpy27/include
Checking for include file webp/demux.h in /usr/local/include
Checking for include file webp/demux.h in /usr/include
Checking for include file webp/demux.h in /usr/include/python2.7
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 3.3.0.dev0
platform     linux2 2.7.5 (default, Nov 20 2015, 02:00:19)
             [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
--------------------------------------------------------------------
--- TKINTER support available
--- JPEG support available
--- OPENJPEG (JPEG2000) support available (2.1)
--- ZLIB (PNG/ZIP) support available
--- LIBTIFF support available
--- FREETYPE2 support available
--- LITTLECMS2 support available
--- WEBP support available
*** WEBPMUX support not available
--------------------------------------------------------------------
To add a missing option, make sure you have the required
library and headers.
See https://pillow.readthedocs.org/en/latest/installation.html#building-from-source

To check the build, run the selftest.py script.
```

**Comment by csucat** [comment]
@wiredfool 
On my centos, the output is:
```
running build_ext
Appending path /home/jdhao/Pillow-master/libImaging
Appending path /usr/lib
Appending path /usr/include
Appending path /usr/lib64
Tkinter found, will check for Tcl/Tk
Checking /py27/Tcl for tk.h
Checking /python27/Tcl for tk.h
Checking /Tcl for tk.h
Checking /Tcl85 for tk.h
Checking /Tcl8.5 for tk.h
Checking /home/jdhao/Pillow-master/Tcl for tk.h
Tcl/tk not found
Appending path /usr/local/lib
Appending path /usr/local/include
Looking for zlib
Checking for include file zlib.h in /home/jdhao/Pillow-master/libImaging
Checking for include file zlib.h in /usr/include
Found zlib.h
Found library z at /usr/lib/libz.so
Looking for jpeg
Checking for include file jpeglib.h in /home/jdhao/Pillow-master/libImaging
Checking for include file jpeglib.h in /usr/include
Found jpeglib.h
Found library jpeg at /usr/lib64/libjpeg.so
Looking for jpeg2000
Checking for openjpeg-#.# in /home/jdhao/Pillow-master/libImaging
Checking for openjpeg-#.# in /usr/include
Found openjpeg.h in /usr/include/openjpeg-2.1
Best openjpeg version (2, 1) so far in /usr/include/openjpeg-2.1
Found openjpeg.h in /usr/include/openjpeg-1.5
Checking for openjpeg-#.# in /usr/local/include
Checking for openjpeg-#.# in /usr/include/python2.7
Found library openjp2 at /usr/lib64/libopenjp2.so
Inserting path /usr/include/openjpeg-2.1
Looking for tiff
Checking for include file tiff.h in /usr/include/openjpeg-2.1
Checking for include file tiff.h in /home/jdhao/Pillow-master/libImaging
Checking for include file tiff.h in /usr/include
Found tiff.h
Found library tiff at /usr/lib64/libtiff.so
Looking for freetype
Found library freetype at /usr/lib64/libfreetype.so
Checking for include file ft2build.h in /usr/include/openjpeg-2.1
Checking for include file ft2build.h in /usr/include/openjpeg-2.1/freetype2
Checking for include file ft2build.h in /home/jdhao/Pillow-master/libImaging
Checking for include file ft2build.h in /home/jdhao/Pillow-master/libImaging/free                          type2
Checking for include file ft2build.h in /usr/include
Found ft2build.h in /usr/include
Inserting path /usr/include/freetype2
Looking for lcms
Checking for include file lcms2.h in /usr/include/freetype2
Checking for include file lcms2.h in /usr/include/openjpeg-2.1
Checking for include file lcms2.h in /home/jdhao/Pillow-master/libImaging
Checking for include file lcms2.h in /usr/include
Found lcms2.h
Found library lcms2 at /usr/lib64/liblcms2.so
Checking for include file tk.h in /usr/include/freetype2
Checking for include file tk.h in /usr/include/openjpeg-2.1
Checking for include file tk.h in /home/jdhao/Pillow-master/libImaging
Checking for include file tk.h in /usr/include
Found tk.h
Looking for TCL
Couldn't find library tcl85 in ['/usr/lib', '/usr/lib64', '/usr/local/lib', '/usr                          /lib64']
Found library tcl8.5 at /usr/lib64/libtcl8.5.so
Looking for TK
Couldn't find library tk85 in ['/usr/lib', '/usr/lib64', '/usr/local/lib', '/usr/                          lib64']
Found library tk8.5 at /usr/lib64/libtk8.5.so
Looking for webp
Checking for include file webp/encode.h in /usr/include/freetype2
Checking for include file webp/encode.h in /usr/include/openjpeg-2.1
Checking for include file webp/encode.h in /home/jdhao/Pillow-master/libImaging
Checking for include file webp/encode.h in /usr/include
Found webp/encode.h
Checking for include file webp/decode.h in /usr/include/freetype2
Checking for include file webp/decode.h in /usr/include/openjpeg-2.1
Checking for include file webp/decode.h in /home/jdhao/Pillow-master/libImaging
Checking for include file webp/decode.h in /usr/include
Found webp/decode.h
Found library webp at /usr/lib64/libwebp.so
Looking for webpmux
Checking for include file webp/mux.h in /usr/include/freetype2
Checking for include file webp/mux.h in /usr/include/openjpeg-2.1
Checking for include file webp/mux.h in /home/jdhao/Pillow-master/libImaging
Checking for include file webp/mux.h in /usr/include
Found webp/mux.h
Checking for include file webp/demux.h in /usr/include/freetype2
Checking for include file webp/demux.h in /usr/include/openjpeg-2.1
Checking for include file webp/demux.h in /home/jdhao/Pillow-master/libImaging
Checking for include file webp/demux.h in /usr/include
Checking for include file webp/demux.h in /usr/local/include
Checking for include file webp/demux.h in /usr/include/python2.7
--------------------------------------------------------------------
PIL SETUP SUMMARY
--------------------------------------------------------------------
version      Pillow 3.3.0.dev0
platform     linux2 2.7.5 (default, Nov 20 2015, 02:00:19)
             [GCC 4.8.5 20150623 (Red Hat 4.8.5-4)]
--------------------------------------------------------------------
--- TKINTER support available
--- JPEG support available
--- OPENJPEG (JPEG2000) support available (2.1)
--- ZLIB (PNG/ZIP) support available
--- LIBTIFF support available
--- FREETYPE2 support available
--- LITTLECMS2 support available
--- WEBP support available
*** WEBPMUX support not available
--------------------------------------------------------------------
To add a missing option, make sure you have the required
library and headers.
See https://pillow.readthedocs.org/en/latest/installation.html#building-from-source

To check the build, run the selftest.py script.
```
The line about the zlib is different, on your CentOS, it is
```Found library z at /usr/lib64/libz.so```
But on my CentOS, it is 
``` Found library z at /usr/lib/libz.so```

I did not do any specific settings,  why the output is different?

**Comment by wiredfool** [comment]
The order that paths are being appended is different: 

Mine:
```
Appending path /home/centos/Pillow/libImaging
Appending path /home/centos/vpy27/lib
Appending path /home/centos/vpy27/include
Appending path /usr/lib64
...
Appending path /usr/local/lib
Appending path /usr/local/include
Appending path /usr/lib
Appending path /usr/include
```

Yours: 
```
Appending path /home/jdhao/Pillow-master/libImaging
Appending path /usr/lib
Appending path /usr/include
Appending path /usr/lib64
...
Appending path /usr/local/lib
Appending path /usr/local/include
```

Your search path is putting /usr/lib in where mine is adding the [virtualenv]/lib directory. So mine is pulling in the 64bit multiarch libs like it's supposed to, before going to the more general locations. I suspect that it's picking that up from the python sysconfig.get_config_var("prefix").
